homeassistant/components/russound_rio/config_flow.py

```diff
@@ -0,0 +1,127 @@
+"""Config flow to configure russound_rio component."""
+
+from __future__ import annotations
+
+import asyncio
+import logging
+from typing import Any
+
+from aiorussound import Russound
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PORT
+from homeassistant.helpers import config_validation as cv
+
+from .const import DOMAIN, RUSSOUND_RIO_EXCEPTIONS, NoPrimaryControllerException
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_NAME): cv.string,
+        vol.Optional(CONF_PORT, default=9621): cv.port,
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def find_primary_controller_metadata(controllers):
+    """Find the mac address of the primary Russound controller."""
+    for controller_id, mac_address, controller_type in controllers:
+        # The integration only cares about the primary controller linked by IP and not any downstream controllers
+        if controller_id == 1:
+            return (mac_address, controller_type)
+    return None
+
+
+class FlowHandler(ConfigFlow, domain=DOMAIN):
+    """Russound RIO configuration flow."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initialized by the user."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            host = user_input[CONF_HOST]
+            name = user_input[CONF_NAME]
+            port = user_input[CONF_PORT]
+
+            try:
+                russ = Russound(self.hass.loop, host, port)
+                async with asyncio.timeout(5):
+                    await russ.connect()
+                    controllers = await russ.enumerate_controllers()
+                    metadata = find_primary_controller_metadata(controllers)
+                    if metadata:
+                        await self.async_set_unique_id(metadata[0])
+                    else:
+                        raise NoPrimaryControllerException
```

Comment:
When do we expect to get this?

Reply:
We will get a NoPrimaryControllerException when there is a connected controller that does not have an ID of one.

To provide context:
The Russound MCA-C5 can be daisy chained (using their proprietary serial connection) up to six total units. On the rear of each unit, there is a dip switch to select the ID of the unit (1-6). Per the unit and API's instructions, only the first logical unit (Unit 1) can be interfaced over IP. In our use case, we treat all the daisy chained controllers to the primary one single device since that's how it's intended by the manufacturer. It's possible for someone to have a single unit on ID 2 that is hooked up to HA over IP. The API doesn't function properly (as expected) in this case so this is caught here to ensure that all controller stack-ups have at least the primary controller present.

This exception may become irrelevant when I pull the logic out of the try-except-else since I was using it as a way to reach the except block.

Reply:
Cleaned this up so that NoPrimaryControllerException is raised directly by the find_primary_controller_metadata method. Leaving this open for discussion / clarifications if there are any.